core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -69,7 +69,7 @@
     this.schema = spec.schema();
     this.nameToField = indexSpecByName(spec);
     this.transformToField = indexSpecByTransform(spec);
-    this.lastAssignedPartitionId = base.lastAssignedPartitionId();
+    this.lastAssignedPartitionId = formatVersion == 1 ? spec.lastAssignedFieldId() : base.lastAssignedPartitionId();
```

Comment:
IIRC, there are multiple places handling it for v1, e.g. https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L345-L350
If we use an iceberg library version with `BaseUpdatePartitionSpec` support, will the base's lastAssignedPartitionId always be set?

Reply:
I think the spec [assumes](https://iceberg.apache.org/spec/#version-2) that `base.lastAssignedPartitionId` is always set for v2 tables.
I checked `PartitionSpecParser` and it looks like `lastAssignedFieldId` is always propagated for v1 tables.

Another pair of eyes to check this would be much appreciated.

Reply:
For v1 table, it might be missing in metadata JSON but the library will still generate it by deriving it from the specs based on the above logic. So I am wondering if there is any case or tests showing that `lastAssignedFieldId` is unset in the loaded table metadata.

Reply:
I see what you are saying. I think you are right. We hit this problem because tests rely on `TestTableOperations` that does not read the metadata file upon committing. While Hadoop and Hive table ops both read the committed files, the question is whether that's part of `TableOperations` contract.

I'll take a look tomorrow.